### PR TITLE
Add basic I/O APIC support.

### PR DESCRIPTION
**Pull request by dlrobertson**
## Summary

Add basic I/O APIC support with the ioapic module in mythril_core.

## TODO

 - [x] Make the `IoApic` more usable
    - [x] Add a `Mutex`
    - [x] Add tests
    - [x] Find the available I/O APICS in `kmain`
 - [x] Make use of the IO APIC

Fixes: #26

### CODE REVIEW

**Comment by dlrobertson** [comment]
@ALSchwalm We're getting the I/O APIC version and ID from the I/O APIC. Is there anything else you'd like me to do in the initial implementation?

```
Loading Mythril
[   0.000000] MYTHRIL-INFO: tsc calibrate diff=3094413 (khz=1802831)
[   0.000018] MYTHRIL-INFO: Allocating from 0x1f58ae8-3ffe0000
[   0.001764] MYTHRIL-INFO: RSDP: revision=1.0 rsdt=0x3ffe15c9
[   0.003675] MYTHRIL-INFO: RSDT: rev=1 length=48 table=0x3ffe15ed
[   0.005703] MYTHRIL-INFO: FACP: rev=1 length=116 table=0x3ffe14c1
[   0.007695] MYTHRIL-INFO: APIC: rev=1 length=128 table=0x3ffe1535
[   0.009740] MYTHRIL-INFO: HPET: rev=1 length=56 table=0x3ffe15b5
[   0.011808] MYTHRIL-INFO: HPET: rev=1 length=56 table=0x3ffe15b5 HPET address=0xfed00000
[   0.014580] MYTHRIL-INFO: IOAPIC: id: 0 version: 0x11 address: 0xfec00000 GSI: 0
```

File: mythril_core/src/ioapic.rs [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+    /// See section 3.2.3 of the I/O APIC specification.
+    pub fn set_arbitration_id(&self, id: u8) -> Result<()> {
+        if id > 15 {
+            Err(Error::InvalidValue(format!(
+                "I/O APIC Arbitration ID `0x{:x}` too large",
+                id
+            )))
+        } else {
+            unsafe {
+                self.write_raw(0x02, (id as u32) << 24);
```
Comment (ALSchwalm) [inline mythril_core/src/ioapic.rs:123]: What is the significance of 0x02 here? Should this be an enum?

File: mythril_core/src/ioapic.rs [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+            }
+            Ok(())
+        }
+    }
+
+    /// Read the raw 64-bit value from the IO Redirect Table Register
+    /// for the given ID.
+    ///
+    /// See section 3.2.4 of the I/O APIC specification.
+    fn read_ioredtbl_raw(&self, id: u8) -> Result<u64> {
```
Comment (ALSchwalm) [inline mythril_core/src/ioapic.rs:133]: I think these `raw` versions should just be unsafe and return a u64, as we don't have any kind of checks in `raw_read` and we don't have any path that can actually return error

File: mythril_core/src/ioapic.rs [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+        match value {
+            0x00 => Ok(DestinationMode::Physical),
+            0x01 => Ok(DestinationMode::Logical),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid destination mode: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
```
Comment (ALSchwalm) [inline mythril_core/src/ioapic.rs:253]: Can we not just use the try_from_primitive stuff we use elsewhere instead of this?

Comment (ALSchwalm) [reply mythril_core/src/ioapic.rs:253]: For pretty much all of these TryFrom's actually.

File: mythril_core/src/ioapic.rs [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+        };
+
+        entry.is_valid()?;
+
+        Ok(entry)
+    }
+
+    /// Perform basic validity checks found in the table from section 3.2.4
+    /// in the I/O APIC specification.
+    fn is_valid(&self) -> Result<()> {
```
Comment (ALSchwalm) [inline mythril_core/src/ioapic.rs:443]: I understand the value of having some context for the error, so we don't want to just return bool for this. But given this signature, I'd like it to be called something like `validate`.

File: mythril_core/src/lib.rs [hunk 5]
```diff
@@ -22,6 +22,8 @@ pub mod device;
 pub mod emulate;
 pub mod error;
 pub mod interrupt;
+/// Support for I/O APIC.
```
Comment (ALSchwalm) [inline mythril_core/src/lib.rs:25]: This seems obvious

File: mythril_core/src/ioapic.rs [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+    /// All addressed local APICs will assume their INIT state. Must
+    /// be edge triggered.
+    INIT = 0b101,
+    /// Deliver the signal to the INTR signal of all processor cores
+    /// listed in destination. Must be edge triggered.
+    ExtINT = 0b111,
+}
+
+impl DeliveryMode {
+    fn level_is_valid(&self) -> bool {
```
Comment (ALSchwalm) [inline mythril_core/src/ioapic.rs:333]: Maybe `valid_for_level_trigger`? Or something else, the current name seems confusing to me.

File: mythril_core/src/ioapic.rs (hunk 1 above)
Comment (dlrobertson) [reply mythril_core/src/ioapic.rs:123]: Ah, yeah it is the ID register. We should have constants for the I/O APIC Register offsets.

File: mythril_core/src/ioapic.rs (hunk 3 above)
Comment (dlrobertson) [reply mythril_core/src/ioapic.rs:253]: We can use `try_from_primitive`, but we'll end up with a similar amount of code in the function where this is used with `.map_err` right?

Comment (dlrobertson) [reply mythril_core/src/ioapic.rs:275]: Posted https://github.com/mythril-hypervisor/mythril/issues/64 after this is implemented I'll use `TryFromPrimitive` here.

**Comment by dlrobertson** [comment]
@ALSchwalm updated

**Comment by ALSchwalm** [comment]
Looks good to me. Merged

**Pull request merged**